### PR TITLE
[#14715][feat] Add Qwen3 embedding model support and /v1/embeddings serving endpoint

### LINKED ISSUE #14715: [Feature]: Add Qwen3-Embedding model support (0.6B/4B/8B) for LLM API encode_only mode
### 🚀 The feature, motivation and pitch

Native support for the [Qwen3-Embedding](https://huggingface.co/Qwen/Qwen3-Embedding-8B) model family (0.6B, 4B, 8B) for use with the LLM API's `encode_only` mode.

### Problem

The Qwen3 embedding models don't work correctly out of the box with `encode_only`. TensorRT-LLM returns logits (post lm_head), whereas the correct output should be the last prompt token's hidden state before the lm_head ([model card's transformers usage](https://huggingface.co/Qwen/Qwen3-Embedding-8B#transformers-usage))

### Proposed support

Add a new architecture/model class, e.g. Qwen3ForEmbedding, for use with `LLM(..., encode_only=True)`.

The model class would:

- Reuse the existing Qwen3 transformer backbone without applying lm_head.
- Return last prompt token's hidden states from `LLM.encode(...)` instead of logits.
- Leave normalization and MRL/custom-dimension slicing to the caller or serving layer.

### Alternatives

_No response_

### Additional context

Models: [Qwen3-Embedding-0.6B](https://huggingface.co/Qwen/Qwen3-Embedding-0.6B), [Qwen3-Embedding-4B](https://huggingface.co/Qwen/Qwen3-Embedding-4B), [Qwen3-Embedding-8B](https://huggingface.co/Qwen/Qwen3-Embedding-8B)

I'm happy to contribute a PR for this feature

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by anurags25**
## Summary by CodeRabbit

* **New Features**
  * Added embedding model support for Qwen3.
  * Added OpenAI-compatible `/v1/embeddings` API endpoint supporting configurable encoding formats, optional dimension slicing, and L2 normalization.

* **Tests**
  * Added integration accuracy tests comparing embedding outputs against reference implementations.

## Description

Resolves #14715 and #10185.

TensorRT-LLM lacks support for serving embedding models via the OpenAI-compatible `/v1/embeddings` API. This PR adds support for [Qwen3 embedding models](https://huggingface.co/Qwen/Qwen3-Embedding-8B) using the PyTorch backend, along with a server endpoint to expose it.

**Approach:** `Qwen3ForEmbedding` uses the generate path (`max_tokens=1`) with `additional_model_outputs` to extract last-token hidden states. This leverages scheduler batching, KV cache, and chunked-prefill; these are needed for production serving of long sequences (up to 32k tokens) that the encode-only path (`llm.encode()`) does not currently support. The support should migrate to the encode path in future once it supports these capabilities.

**Changes:**
- `Qwen3ForEmbedding` model class extending `Qwen3ForCausalLM` with last-token pooling
- `Qwen3EmbeddingHfWeightMapper` handling both prefixed and unprefixed HF checkpoint layouts
- `EmbeddingRequest/Data/Response` protocol models (OpenAI-compatible)
- `/v1/embeddings` server handler with base64/float encoding, MRL dimension truncation, and L2 normalization
- Integration test comparing TRT-LLM output against HuggingFace baseline

**Design decisions:**
- **Generate path over encode path**: The encode path runs a single synchronous forward pass without scheduling, chunked prefill, or concurrent request handling. For production serving of long-context embedding models, the generate path is required.
- **Output name via env var**: `TRTLLM_EMBEDDING_OUTPUT_NAME` configures the `additional_model_outputs` key, keeping the endpoint model-agnostic pending a proper model-level config mechanism.

**Caveats:**
- **KV cache reuse must be disabled**: When `enable_block_reuse=True`, cached prefixes skip the forward pass so `additional_model_outputs` are not computed for reused tokens. Set `enable_block_reuse: false` for embedding workloads.
- **`encode_only` returns logits, not embeddings**: Using `LLM(encode_only=True)` with this model returns token logits (standard decoder behavior). Embeddings require the generate path with `additional_model_outputs`.

**Usage (LLM API):**

```python
from tensorrt_llm import LLM, SamplingParams
from tensorrt_llm.llmapi import KvCacheConfig
import torch.nn.functional as F

llm = LLM(
    model="Qwen/Qwen3-Embedding-0.6B",
    backend="pytorch",
    max_input_len=4096,
    max_seq_len=4097,
    enable_chunked_prefill=True,
    kv_cache_config=KvCacheConfig(enable_block_reuse=False),
    model_kwargs={
        "architectures": ["Qwen3ForEmbedding"],
        "tie_word_embeddings": True,
    },
)

outputs = llm.generate(
    ["hello world"],
    SamplingParams(
        max_tokens=1, temperature=0.0, detokenize=False,
        additional_model_outputs=["last_token_hidden_state"],
    ),
)
hidden_state = outputs[0].outputs[0].additional_generation_outputs["last_token_hidden_state"]
embedding = F.normalize(hidden_state.float(), p=2, dim=-1)
```

**Usage (trtllm-serve):**

    TRTLLM_EMBEDDING_OUTPUT_NAME=last_token_hidden_state \
      trtllm-serve ./qwen3-embedding-8b/ --config config.yaml

    # config.yaml
    model_kwargs:
      architectures: ["Qwen3ForEmbedding"]
      tie_word_embeddings: true
    kv_cache_config:
      enable_block_reuse: false

## Test Coverage

- `tests/integration/defs/accuracy/test_llm_api_pytorch_embedding.py` — accuracy test comparing TRT-LLM embeddings against HuggingFace baseline using cosine similarity (>0.99 bf16) and retrieval score alignment (<0.02 max diff).

Validated manually on H200 with Qwen3-Embedding-8B (bf16 and FP8), with chunked prefill and perf stats enabled.

## PR Checklist

- [x] Please check this after reviewing the above items as appropriate for this PR.

### CODE REVIEW

File: tensorrt_llm/serve/openai_server.py [hunk 1]
```diff
@@ -1379,6 +1387,85 @@ async def create_mm_embedding_response(promise: RequestOutput):
...
+        using the generate path for scheduler batching and chunked-prefill
+        support. This should migrate to the encode path once it supports
+        equivalent capabilities. The output tensor name is configured
+        via the TRTLLM_EMBEDDING_OUTPUT_NAME env var.
+        """
+        try:
+            texts = ([request.input]
+                     if isinstance(request.input, str) else request.input)
+
+            output_name = os.environ["TRTLLM_EMBEDDING_OUTPUT_NAME"]
```
[the remark this answers is not in the document]
Comment (anurags25) [reply tensorrt_llm/serve/openai_server.py]: Fixed.

**Review by jieli-matrix — APPROVED** [review]

**Comment by anurags25** [comment]
@jieli-matrix Thank you for the review! Can you help merge the PR as well?

File: tests/integration/defs/accuracy/test_llm_api_pytorch_embedding.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        x = out.outputs[0].additional_generation_outputs[OUTPUT_NAME]
+        if x.ndim == 3:
+            x = x[0, 0]
+        elif x.ndim == 2:
+            x = x[0]
+        embeddings.append(F.normalize(x.float(), p=2, dim=-1).cpu())
+    return torch.stack(embeddings)
+
+
+class TestQwen3Embedding8B(LlmapiAccuracyTestHarness):
```
Comment (hchings) [inline tests/integration/defs/accuracy/test_llm_api_pytorch_embedding.py:143]: Can you add this test file in ` tests/integration/test_lists/test-db/l0_a100.yml` for CI to include it (around L39)?

File: tensorrt_llm/serve/openai_server.py [hunk 3]
```diff
@@ -1382,6 +1390,92 @@ async def create_mm_embedding_response(promise: RequestOutput):
...
+        using the generate path for scheduler batching and chunked-prefill
+        support. This should migrate to the encode path once it supports
+        equivalent capabilities. The output tensor name is configured
+        via the TRTLLM_EMBEDDING_OUTPUT_NAME env var.
+        """
+        try:
+            texts = ([request.input]
+                     if isinstance(request.input, str) else request.input)
+
+            output_name = os.environ.get("TRTLLM_EMBEDDING_OUTPUT_NAME")
```
Comment (hchings) [inline tensorrt_llm/serve/openai_server.py:1566]: Is this design aligned w/ vllm's? i.e., A env var must be set for this endpoint to work.

Also a nit - can we update `docs/source/commands/trtllm-serve/trtllm-serve.rst`?

Comment (anurags25) [reply tensorrt_llm/serve/openai_server.py:1566]: Added docs.

vLLM doesn't use an env var for this, their design would align with `LLM.encode()` call. The env var here is an interim solution because we're using the generate path to get batching and chunked prefill support that the encode path doesn't have yet. Once encode supports those, /v1/embeddings should switch over and the `additional_model_outputs` extraction can be dropped (and the env var with it).

I can also add a brief comment in the code explaining this if that'd be helpful.

Comment (hchings) [reply tensorrt_llm/serve/openai_server.py:1566]: Yes, a brief code comment to show that this env var is a workaround will be helpful. Thanks!

**Review by hchings — APPROVED** [review]

**Review by nvrohanv — APPROVED** [review]

**Comment by hchings** [comment]
CI failed at not having `https://huggingface.co/Qwen/Qwen3-Embedding-8B` llm-models. I've asked CI team to upload.

File: tests/integration/defs/accuracy/test_llm_api_pytorch_embedding.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    "It gives weight to physical objects and is responsible for the "
+    "movement of planets around the sun.",
+]
+
+MAX_INPUT_LEN = 4096
+OUTPUT_NAME = "last_token_hidden_state"
+
+EMBEDDING_MODELS = [
+    pytest.param(
+        f"{llm_models_root()}/qwen3-embedding-8b",
```
Comment (hchings) [inline tests/integration/defs/accuracy/test_llm_api_pytorch_embedding.py]: Hi can you change this to `f"{llm_models_root()}/Qwen3/Qwen3-Embedding-8B"` to match the model location in CI otherwise it'll fail?

**Comment by anurags25** [comment]
Closing this PR as proper support with encoder only path was added in #15424; appreciate folks taking the time to review!
